### PR TITLE
Fix hot restart bug due to passed an fd without bind

### PR DESCRIPTION
**Pull request by zyfjeff**
When the 15001 listener appears at the same time, one of them will bind the socket and listening, and the other will not bind. In this case, the hot restart gets the socket of the previous process and may get the socket that is not listening.

Description: Fix hot restart bug due to passed an fd without bind
Risk Level: low
Testing:  Existing tests and new tests
Docs Changes: N/A
Release Notes: N/A
[Optional Fixes #Issue] #8427

### CODE REVIEW

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -166,6 +166,11 @@ Network::SocketSharedPtr ProdListenerComponentFactory::createListenSocket(
                                  ? Network::Utility::TCP_SCHEME
                                  : Network::Utility::UDP_SCHEME;
   const std::string addr = absl::StrCat(scheme, address->asString());
+
+  if (!bind_to_port && socket_type == Network::Address::SocketType::Stream) {
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: Is this needed? Won't this just return -1 below once you check for not binding in the parent?

File: source/server/hot_restarting_parent.cc [hunk 2]
```diff
@@ -95,6 +95,11 @@ HotRestartingParent::Internal::getListenSocketsForChild(const HotRestartMessage:
   Network::Address::InstanceConstSharedPtr addr =
       Network::Utility::resolveUrl(request.pass_listen_socket().address());
   for (const auto& listener : server_->listenerManager().listeners()) {
+    if (listener.get().socket().socketType() == Network::Address::SocketType::Stream &&
```
Comment (mattklein123) [inline source/server/hot_restarting_parent.cc]: Pedantically, do we need the stream type check? Isn't just checking for bind to port sufficient?

Comment (zyfjeff) [reply source/server/hot_restarting_parent.cc]: Yes, you are right, I understand it wrong, udp socket has no bind, there is no way to handle traffic, so you also need to consider, here is only need to consider whether you can bind

File: source/server/listener_manager_impl.cc (hunk 1 above)
Comment (zyfjeff) [reply source/server/listener_manager_impl.cc]: Here is the need to determine whether the bind, if there is no bind socket to the parent process to obtain the socket, may get the already bound socket, in addition, there is no way to handle the traffic without the bind socket, you can not go to the parent process to obtain the socket, directly Create it.

**Comment by mattklein123** [comment]
@zyfjeff friendly request to please not force push. It makes reviewing more difficult. Thank you!

File: source/server/listener_manager_impl.cc [hunk 3]
```diff
@@ -166,16 +166,20 @@ Network::SocketSharedPtr ProdListenerComponentFactory::createListenSocket(
...
-  const int fd = server_.hotRestart().duplicateParentListenSocket(addr);
-  if (fd != -1) {
-    ENVOY_LOG(debug, "obtained socket for address {} from parent", addr);
-    Network::IoHandlePtr io_handle = std::make_unique<Network::IoSocketHandleImpl>(fd);
-    if (socket_type == Network::Address::SocketType::Stream) {
-      return std::make_shared<Network::TcpListenSocket>(std::move(io_handle), address, options);
-    } else {
-      return std::make_shared<Network::UdpListenSocket>(std::move(io_handle), address, options);
+
+  if (bind_to_port) {
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: Can we add a listener manager test for this case also?

Comment (zyfjeff) [reply source/server/listener_manager_impl.cc]: Yes, the idea of ​​my test case is to test when the bind_to_port is equal to false, not to get the socket from the hot restart, but to create the socket directly.

File: test/server/listener_manager_impl_test.cc [hunk 4]
```diff
@@ -915,6 +916,45 @@ name: foo
...
+  auto syscall_result = os_syscall.socket(AF_INET, SOCK_STREAM, 0);
+  ASSERT_GE(syscall_result.rc_, 0);
+  ListenerHandle* listener_foo = expectListenerCreate(true, true);
+  EXPECT_CALL(listener_factory_, createListenSocket(_, _, _, false))
+      .WillOnce(
+          Invoke([this, &syscall_result](const Network::Address::InstanceConstSharedPtr& address,
+                                         Network::Address::SocketType socket_type,
+                                         const Network::Socket::OptionsSharedPtr& options,
+                                         bool bind_to_port) -> Network::SocketSharedPtr {
+            // When bind_to_port is equal to false, create socket fd directly, and do not get socket
```
Comment (mattklein123) [inline test/server/listener_manager_impl_test.cc]: Can you add an explicit expectation with `Times(0)` to the mock hot restarter to make sure we never call it?

File: test/server/listener_manager_impl_test.h [hunk 5]
```diff
@@ -37,7 +37,7 @@ class ListenerHandle {
 
 class ListenerManagerImplTest : public testing::Test {
 protected:
-  ListenerManagerImplTest() : api_(Api::createApiForTest()) {
+  ListenerManagerImplTest() : api_(Api::createApiForTest()), real_listener_factory_(server_) {
```
Comment (mattklein123) [inline test/server/listener_manager_impl_test.h]: Can you just define this in the test where you use it for now as we don't use it anywhere else?

**Pull request merged**